### PR TITLE
Extract wpt-pr-results code into lib/results.js

### PR DESCRIPTION
**Pull request by stephenmcgruer**

### CODE REVIEW

File: lib/results.js [hunk 1]
```diff
@@ -0,0 +1,157 @@
+'use strict';
+
+/**
+ * Utility functions for interacting with WPT run results encoded in a git
+ * repository (wpt-pr-results).
+ *
+ * The wpt-pr-results repository stores results from WPT runs as individual
+ * orphan commits, each pointed to by a tag. A given commit (aka run) stores the
+ * results in expanded directory form, where each WPT test has a results JSON
+ * file stored at /root/path/to/test/test_name.json.
```
Comment (foolip) [inline lib/results.js]: `test_name` is actually the base64-encoded test name now I think.

File: lib/results.js [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+ * When a run is loaded into memory, we store it as a tree, where each node
+ * represents a directory. A node has a (possibly empty) map of directory name
+ * to child node and a (possibly empty) map of test name (for tests in the
+ * node's directory) to results JSON.
+ *
+ * The results JSON for a particular test looks something like:
+ *
+ *   {status: OK, subtests: [{ name: "Foo", status: "PASS" }, ...]}
+ *
+ * The 'subtests' array may be missing, if the test is a single page test[0]. In
```
Comment (foolip) [inline lib/results.js]: The more common case where it will be missing is reftests.

File: lib/results.js [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    const parts = tag.toString().split('/');
+    return Number(parts[3]);
+  }));
+}
+
+// Walks an input tree in depth-first order, calling the visitor function on
+// each test in the tree. The visitor function should be of the form:
+//   visitor(path, test_name, test_results)
+//
+// Where test_results is a json dictionary as described in the module
```
Comment (foolip) [inline lib/results.js]:
```suggestion
// Where test_results is an object as described in the module
```

JSON is a string, the actual value is an object :)

File: lib/results.js (hunk 1 above)
Comment (stephenmcgruer) [reply lib/results.js]: I'm not sure I understand what you mean. For example:

```
$ git show run/375980002/results
commit 678a889b18b356ff377204fa894b4adfafdc8427 (tag: run/375980002/results)
Author: autofoolip <auto@foolip.org>
Date:   Tue Nov 19 02:27:40 2019 +0000

    https://storage.googleapis.com/wptd-results/ee893908a206f562a35f202fce67e667bc19f603/chrome-78.0.3904.108-linux-18.04-1ae3cb70cb/report.json

diff --git a/2dcontext/building-paths/canvas_complexshapes_arcto_001.htm.json b/2dcontext/building-paths/canvas_complexshapes_arcto_001.htm.json
new file mode 100644
index 0000000000..b8653da867
--- /dev/null
+++ b/2dcontext/building-paths/canvas_complexshapes_arcto_001.htm.json
@@ -0,0 +1 @@
+{"status":"PASS"}
\ No newline at end of file
...
```

That doesn't look base64 encoded to me?

Comment (foolip) [reply lib/results.js]: I misremembered. base64 was one fix I considered when `?` started showing up in test names, but I went with URL encoding instead: https://github.com/foolip/ad-hoc-wpt-results-analysis/commit/44ed0f9eeb28428f03b85e650f60e29e2b37436f

**Pull request merged**